// @microsoft/mixed-reality-extension-sdk/built/types/runtime/mediaInstance.js
"use strict";
/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const v4_1 = __importDefault(require("uuid/v4"));
const __1 = require("../..");
const log_1 = require("../../log");
/**
 * A MediaInstance represents an instance managing the playback of a sound or video stream,
 * i.e. it plays an asset that was preloaded in an asset container
 */
class MediaInstance {
    constructor(actor, mediaAssetId) {
        this.id = v4_1.default();
        this.actor = actor;
        this.mediaAssetId = mediaAssetId;
    }
    /**
     * @hidden
     */
    start(options) {
        this.actor.context.internal.lookupAsset(this.mediaAssetId).created.then(() => {
            this.actor.context.internal.setMediaState(this, __1.MediaCommand.Start, options, this.mediaAssetId);
        }).catch(reason => {
            log_1.log.error('app', `Start failed ${this.actor.id}. ${(reason || '').toString()}`.trim());
        });
        return this;
    }
    /**
     * Updates the state of the active media
     * @param options Adjustments to pitch and volume, and other characteristics.
     */
    setState(options) {
        this.actor.context.internal.lookupAsset(this.mediaAssetId).created.then(() => {
            this.actor.context.internal.setMediaState(this, __1.MediaCommand.Update, options);
        }).catch((reason) => {
            log_1.log.error('app', `SetState failed ${this.actor.id}. ${(reason || '').toString()}`.trim());
        });
    }
    /**
     * Pause the media playback
     */
    pause() {
        this.setState({ paused: true });
    }
    /**
     * Unpause the media playback
     */
    resume() {
        this.setState({ paused: false });
    }
    /**
     * Finish the media playback and destroy the instance.
     */
    stop() {
        this.actor.context.internal.lookupAsset(this.mediaAssetId).created.then(() => {
            this.actor.context.internal.setMediaState(this, __1.MediaCommand.Stop);
        }).catch((reason) => {
            log_1.log.error('app', `Stop failed ${this.actor.id}. ${(reason || '').toString()}`.trim());
        });
    }
}
exports.MediaInstance = MediaInstance;
