// @microsoft/mixed-reality-extension-sdk/built/types/internal/user.js
"use strict";
/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @hidden
 */
class InternalUser {
    constructor(user, context) {
        this.user = user;
        this.context = context;
        this.observing = true;
    }
    getPatchAndReset() {
        const patch = this.patch;
        if (patch) {
            patch.id = this.user.id;
            delete this.patch;
        }
        return patch;
    }
    prompt(text, acceptInput) {
        const payload = {
            type: 'show-dialog',
            userId: this.user.id,
            text,
            acceptInput
        };
        return new Promise((resolve, reject) => {
            this.context.sendPayload(payload, { resolve, reject });
        })
            .then(response => {
            if (response.failureMessage) {
                return Promise.reject(response.failureMessage);
            }
            else {
                return Promise.resolve({
                    submitted: response.submitted,
                    text: response.text
                });
            }
        });
    }
}
exports.InternalUser = InternalUser;
